spotify_confidence/analysis/frequentist/confidence_computers/generic_computer.py

```diff
@@ -727,106 +595,257 @@ def _get_num_comparisons(self, df: DataFrame, correction_method: str, groupby: I
-        level 1 and level 2 given a targeted MDE.
-
-        Args:
-            level_1 (str, tuple of str): Name of first level.
-            level_2 (str, tuple of str): Name of second level.
-            mde (float): Absolute minimal detectable effect size.
-            alpha (float): Type I error rate, cutoff value for determining
-                statistical significance.
-            groupby (str): Name of column.
-                If specified, will return the difference for each level
-                of the grouped dimension.
-
-        Returns:
-            Pandas DataFrame with the following columns:
-            - level_1: Name of level 1.
-            - level_2: Name of level 2.
-            - power: 1 - B, where B is the likelihood of a Type II (false
-                negative) error.
-
-        """
-        groupby = listify(groupby)
-        level_columns = get_remaning_groups(self._all_group_columns, groupby)
-        return (
-            self._compute_differences(
-                level_columns,
-                [(level_1, level_2)],
-                True,
-                groupby,
-                level_as_reference=True,
-                nims=None,
-                final_expected_sample_size_column=None,
-                mde_column=None,
-            )  # TODO: IS this right?
-            .pipe(lambda df: df if groupby == [] else df.set_index(groupby))
-            .assign(achieved_power=lambda df: df.apply(self._achieved_power, mde=mde, alpha=alpha, axis=1))
-        )[["level_1", "level_2", "achieved_power"]]
-
-    def _point_estimate(self, df: DataFrame) -> Series:
-        return df.apply(lambda row: self._confidence_computers[row[self._method_column]]._point_estimate(row), axis=1)
 
-    def _variance(self, df: DataFrame) -> Series:
-        return df.apply(lambda row: self._confidence_computers[row[self._method_column]]._variance(row), axis=1)
+def add_nim_input_columns_from_tuple_or_dict(df, nims: NIM_TYPE, mde_column: str) -> DataFrame:
+    if type(nims) is tuple:
+        return df.assign(**{NIM_INPUT_COLUMN_NAME: nims[0]}).assign(**{PREFERRED_DIRECTION_INPUT_NAME: nims[1]})
+    elif type(nims) is dict:
+        nim_values = {key: value[0] for key, value in nims.items()}
+        nim_preferences = {key: value[1] for key, value in nims.items()}
+        return df.assign(**{NIM_INPUT_COLUMN_NAME: lambda df: df.index.to_series().map(nim_values)}).assign(
+            **{PREFERRED_DIRECTION_INPUT_NAME: lambda df: df.index.to_series().map(nim_preferences)}
+        )
+    elif nims is None:
+        return df.assign(**{NIM_INPUT_COLUMN_NAME: None}).assign(
+            **{
+                PREFERRED_DIRECTION_INPUT_NAME: None
+                if PREFERRED_DIRECTION_INPUT_NAME not in df or mde_column is None
+                else df[PREFERRED_DIRECTION_INPUT_NAME]
+            }
+        )
+    else:
+        return df
+
+
+def add_nims_and_mdes(df: DataFrame, mde_column: str) -> DataFrame:
```

Comment:
what is that requires mde to have an argument but not nim?

Reply:
Yeah, this is a bit weird and something to change at some point. For NIMs we have default columns names, defined as constants, that you have to use, while for the MDE, the user passes in the column name. I think we should use the latter for NIMs as well, but it's not something we need to fix now I think.